<?xml version="1.0"?>
<robot name="gimbal_demo" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:macro name="inertial_box" params="mass x y z">
    <inertial>
      <mass value="${mass}"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="${mass/12.0*(y*y + z*z)}"
               ixy="0.0"
               ixz="0.0"
               iyy="${mass/12.0*(x*x + z*z)}"
               iyz="0.0"
               izz="${mass/12.0*(x*x + y*y)}"/>
    </inertial>
  </xacro:macro>

  <!-- base_link -->
  <link name="base_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.2 0.2 0.05"/>
      </geometry>
      <material name="grey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.2 0.2 0.05"/>
      </geometry>
    </collision>
    <xacro:inertial_box mass="1.0" x="0.2" y="0.2" z="0.05"/>
  </link>

  <gazebo>
      <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
          <parameters>$(find gimbal_demo)/config/controller.yaml</parameters>
      </plugin>
  </gazebo>

  <!-- yaw_link -->
  <link name="gimbal_yaw_link">
    <visual>
      <geometry>
        <cylinder length="0.05" radius="0.05"/>
      </geometry>
      <material name="blue">
        <color rgba="0.2 0.2 0.9 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.05" radius="0.05"/>
      </geometry>
    </collision>
    <xacro:inertial_box mass="0.5" x="0.1" y="0.1" z="0.05"/>
  </link>

  <!-- pitch_link -->
  <link name="gimbal_pitch_link">
    <visual>
      <origin xyz="0.0 0.0 0.0" rpy="${pi/2} ${pi/2} 0.0"/>
      <geometry>
        <cylinder length="0.05" radius="0.04"/>
      </geometry>
      <material name="green">
        <color rgba="0.2 0.9 0.2 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.05" radius="0.04"/>
      </geometry>
    </collision>
    <xacro:inertial_box mass="0.4" x="0.08" y="0.08" z="0.05"/>
  </link>

  <!-- camera_link -->
  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="black">
        <color rgba="0.1 0.1 0.1 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>
    <xacro:inertial_box mass="0.2" x="0.05" y="0.05" z="0.05"/>
  </link>

  <!-- camera_optical_frame -->
  <link name="camera_optical_frame">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.01"/>
      </geometry>
      <material name="yellow">
        <color rgba="1.0 1.0 0.0 1.0"/>
      </material>
    </visual>
  </link>

  <!-- yaw joint -->
  <joint name="yaw_joint" type="revolute">
    <parent link="base_link"/>
    <child link="gimbal_yaw_link"/>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.5" upper="2.5" effort="5.0" velocity="2.0"/>
  </joint>

  <!-- pitch joint -->
  <joint name="pitch_joint" type="revolute">
    <parent link="gimbal_yaw_link"/>
    <child link="gimbal_pitch_link"/>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.7" friction="0.1"/>
    <limit lower="-1.57" upper="0.5" effort="5.0" velocity="2.0"/>
  </joint>

  <!-- camera fixed -->
  <joint name="camera_fixed_joint" type="fixed">
    <parent link="gimbal_pitch_link"/>
    <child link="camera_link"/>
    <origin xyz="0.05 0 0" rpy="0 0 0"/>
  </joint>

  <!-- camera optical frame -->
  <joint name="camera_optical_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_optical_frame"/>
    <!-- ROS 카메라 표준 REP-103 optical frame 변환 -->
    <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
  </joint>

  <!-- Gazebo 카메라 센서 -->
  <gazebo reference="camera_link">
    <sensor type="camera" name="gimbal_camera">
      <always_on>true</always_on>
      <update_rate>30.0</update_rate>

      <camera name="gimbal_cam">
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.05</near>
          <far>50.0</far>
        </clip>
      </camera>
      <plugin name="gazebo_ros_camera" filename="libgazebo_ros_camera.so">
        <ros>
          <namespace>/gimbal</namespace>
          <argument>image:=image_raw</argument>
          <argument>camera_info:=camera_info</argument>
        </ros>
        <frame_name>camera_optical_frame</frame_name>
      </plugin>
      <!-- <topic>camera/image_raw</topic> -->
    </sensor>
  </gazebo>



  <ros2_control name="GimbalSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>

    <joint name="yaw_joint">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>

    <joint name="pitch_joint">
      <command_interface name="position"/>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>
</robot>
